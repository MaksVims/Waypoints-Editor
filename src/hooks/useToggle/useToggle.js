import React, {useCallback, useState} from "react";

export function useToggle(defaultValue = false) {
  const [value, setValue] = useState(defaultValue)

  const toggle = useCallback((updateValue) => {
    setValue(value => {
      return typeof updateValue === 'boolean' ? updateValue : !value
    })
  }, [])

  return [value, toggle]
}