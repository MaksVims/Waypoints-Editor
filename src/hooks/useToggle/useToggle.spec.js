import {renderHook} from "@testing-library/react-hooks";
import {useToggle} from "./useToggle";
import {act} from "react-test-renderer";

describe(">>> HOOK --- USE_TOGGLE", function () {

  it('--- should hook return correct values', function () {
    let initValue = true
    const {result} = renderHook(() => useToggle(() => initValue))
    const [value, toggle] = result.current
    expect(value).toEqual(initValue)
    expect(typeof toggle === 'function').toBe(true)
  });

  it('--- should hook return false without init value', function () {
    const {result} = renderHook(() => useToggle())
    const [value] = result.current
    expect(value).toBe(false)
  });

  it('--- should call toggle function without args return inversion boolean value', function () {
    const {result} = renderHook(() => useToggle(false))
    act(() => {
      result.current[1]()
    })
    expect(result.current[0]).toBe(true)
  })

  it('--- should call toggle function with NOT BOOLEAN args return inversion current value', function () {
    const {result} = renderHook(() => useToggle(false))
    act(() => {
      result.current[1]('hello')
    })
    expect(result.current[0]).toBe(true)
  })

  it('--- should return function immutable for rerender', function () {
    const {result, rerender} = renderHook(() => useToggle(false))
    const [value, toggle] = result.current
    rerender()
    expect(toggle).toEqual(result.current[1])
  });

  it('--- equal to snapshot ', function () {
    const result = renderHook(() => useToggle(false))
    expect(result).toMatchSnapshot()
  });
})
