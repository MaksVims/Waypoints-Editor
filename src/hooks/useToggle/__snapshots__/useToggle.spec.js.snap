// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> HOOK --- USE_TOGGLE --- equal to snapshot  1`] = `
Object {
  "rerender": [Function],
  "result": Object {
    "all": Array [
      Array [
        false,
        [Function],
      ],
    ],
    "current": Array [
      false,
      [Function],
    ],
    "error": undefined,
  },
  "unmount": [Function],
  "waitFor": [Function],
  "waitForNextUpdate": [Function],
  "waitForValueToChange": [Function],
}
`;
